File: Data Structure and Algorithm/Assignment 06/Program 4/prog4.py
# 4. Write a Python program to calculate factorial with the help of stack.

class stack:
    def __init__(self,capacity):
        self.stack = []
        self.capacity = capacity
    
    def isEmpty(self):
        if len(self.stack) == 0:
            return 1
        return 0

    def push(self, val):
        if len(self.stack) == self.capacity:
            print("Stack Overflow")
        else:
            self.stack.append(val)
    
    def pop(self):
        if len(self.stack) == 0:
            print("Stack Underflow")
        else:
            return self.stack.pop()
    
def factorial(num):
    s = stack(num+1)
    while num >= 1:
        s.push(num)
        num -=1
    fact = 1
    while not s.isEmpty():
        fact *= s.pop()
    return fact


if __name__ == "__main__":
    num = int(input("Enter the number :"))
    print(factorial(num))

File: Data Structure and Algorithm/Assignment 06/Program 4/prog4.c
// 4. Write a C program to calculate factorial with the help of stack.
#include<stdio.h>
#include<stdlib.h>

typedef struct Stack{
    int top;
    int size;
    int *arr;
}stack;

int isFull(stack *ptr){
    if(ptr->top == ptr->size-1)
        return 1;
    return 0;
}

int isEmpty(stack *ptr){
    if(ptr->top == -1)
        return 1;
    return 0;
}

void push(stack *ptr, int data){
    if(isFull(ptr)){
        printf("Stack Overflow");
        return;
    }
    else{
        ptr->top ++;
        ptr->arr[ptr->top] = data;
    }
}

int pop(stack *ptr){
    int item;
    if(isEmpty(ptr)){
        printf("Stack Underflow");
        return 0;
    }
    else{
        item = ptr->arr[ptr->top];
        ptr->top --;
    }
    return item;
}


int factorial(stack *sp,int num){
    sp = (stack*)malloc(sizeof(stack));
    sp->top = -1;
    sp->size = num;
    sp->arr = (int*)malloc(sp->size*sizeof(int));
    while(num>= 1){
        push(sp,num);
        num -= 1;
    }
    int fact =1;
    while (!isEmpty(sp)){
        int val = pop(sp);
        fact *= val;
    }
    return fact;

}

int main(){
    stack *sp;
    int num;
    printf("Enter the number: ");
    scanf("%d",&num);
    printf("%d",factorial(sp, num));
    return 0;
}